megamek/src/megamek/common/BoardIndex.java

```diff
@@ -0,0 +1,154 @@
+        *  or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
+        *  for more details.
+        */
+
+import megamek.MegaMek;
+
+import java.io.BufferedReader;
+import java.io.File;
+import java.io.FileReader;
+import java.io.FilenameFilter;
+import java.io.IOException;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Locale;
+
+// This is to create an index of the maps available, and allow for querying
+public class BoardIndex {
+
+    private static final String SEPARATOR_COMMA = ",";
+    private static final String FILE_SUFFIX_BOARD = ".board";
+    
+    private List<String> allDirs = new ArrayList<String>();
+    private List<String[]> index = new ArrayList<String[]>();
+    private String indexDir = "";
+    
+    // Constructor, stores the custom map directory, finds the directories in that location, and runs the indexing
+    public BoardIndex(String mapDir) {
+        indexDir = mapDir;
+        indexDirs(mapDir);
+        runIndex();
+    }
+    
+    public static void main () {
+        
+    }
+    
+    // Returns the index array list.
+    public List<String[]> getIndex() { return index; }
+    
+    // Returns the directory list
+    public List<String> getDirs() { return allDirs; }
+
+    // Add a directory to the directory list, after checking if it already exists in the list
+    public void addDir(String dir) { 
+        if (!allDirs.contains(dir)) {
+            allDirs.add(dir);
+        } else {
+            MegaMek.getLogger().info("Did not add directory, as it already exists");
+        }
+    } 
+    
+    // Get a random board of specified width and height.
+    public String getRandom(int width, int height) {
+        List<Integer> randMap = new ArrayList<Integer>();
+
+        for (int i = 0; i < index.size(); i++) {
+            String[] board = index.get(i);
+            if ((Integer.parseInt(board[1]) == width) && (Integer.parseInt(board[2])==height)) {
+                randMap.add(i);
+            }
+        }
+
+        String randomMap = index.get(randMap.get(Compute.randomInt(randMap.size())))[4] + FILE_SUFFIX_BOARD;
+        return randomMap;
+    }
+    
+    // Clear the directory list
+    public void clearDirs() { allDirs.clear(); }
+        
+    // Clear and re-run the index. Useful if maps have been added or directories added
+    public void reIndex() {
+        index.clear();
+        runIndex();
+    }
+    
+    // Index the directories. This calls addDir to add them and check for duplicates.
+    private void indexDirs(String mapDir) {
+ 
+        addDir("");
+
+        MegaMek.getLogger().info("Begin map directory index" +LocalDateTime.now());
+        
+        if (mapDir != null) {
+            for (String customDir : mapDir.split(SEPARATOR_COMMA, -1)) {
+                addDir(customDir);
+            }
+        }
+        File dirList = new File(Configuration.boardsDir(), "");
```

Comment:
Why an empty string for the child?

Reply:
it is to retrieve the directory listing. wasn't sure if I could pass without it. Normally that is where the file goes. All examples I found used "".